applications/serial_lte_modem/src/gnss/slm_at_gnss.c

```diff
@@ -733,44 +733,41 @@ static void on_cloud_evt_disconnected(void)
 	at_monitor_pause(&ncell_meas);
 }
 
-static void on_cloud_evt_data_received(const struct nrf_cloud_data *const data)
+static void on_cloud_evt_cell_pos_data_received(const struct nrf_cloud_data *const data)
 {
-	int err = -EAGAIN;
+	if (run_type != RUN_TYPE_CELL_POS) {
+		LOG_WRN("Not expecting cell pos data during run_type: %d", run_type);
+		return;
+	}
 
-	if (run_type == RUN_TYPE_AGPS) {
-		err = nrf_cloud_agps_process(data->ptr, data->len);
-	} else if (run_type == RUN_TYPE_PGPS) {
-		err = nrf_cloud_pgps_process(data->ptr, data->len);
-	} else if (run_type == RUN_TYPE_CELL_POS) {
-		struct nrf_cloud_cell_pos_result result;
-
-		err = nrf_cloud_cell_pos_process(data->ptr, &result);
-		if (err == 0) {
-			if (ttft_start != 0) {
-				LOG_INF("TTFF %ds", (int)k_uptime_delta(&ttft_start)/1000);
-				ttft_start = 0;
-			}
-			sprintf(rsp_buf, "\r\n#XCELLPOS: %d,%lf,%lf,%d\r\n",
-				result.type, result.lat, result.lon, result.unc);
-			rsp_send(rsp_buf, strlen(rsp_buf));
-			run_type = RUN_TYPE_NONE;
-		} else if (err == 1) {
-			LOG_WRN("No position found");
-		} else if (err == -EFAULT) {
-			LOG_ERR("Unable to determine location from cell data, error: %d",
-				result.err);
+	int err;
+	struct nrf_cloud_cell_pos_result result;
+
+	err = nrf_cloud_cell_pos_process(data->ptr, &result);
+	if (err == 0) {
+		if (ttft_start != 0) {
+			LOG_INF("TTFF %ds", (int)k_uptime_delta(&ttft_start)/1000);
+			ttft_start = 0;
 		}
+		sprintf(rsp_buf, "\r\n#XCELLPOS: %d,%lf,%lf,%d\r\n",
+			result.type, result.lat, result.lon, result.unc);
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		run_type = RUN_TYPE_NONE;
+	} else if (err == 1) {
+		LOG_WRN("No position found");
+	} else if (err == -EFAULT) {
+		LOG_ERR("Unable to determine location from cell data, error: %d",
+			result.err);
 	} else {
-		if (nrf_cloud_ready) {
-			sprintf(rsp_buf, "\r\n#XNRFCLOUD: %s\r\n", (char *)data->ptr);
-			rsp_send(rsp_buf, strlen(rsp_buf));
-			err = 0;
-		}
+		LOG_ERR("Failed to process cellular location response, error: %d",
+			err);
 	}
+}
 
-	if (err < 0) {
-		LOG_WRN("Unable to process data, error: %d run_type: %d", err, run_type);
-	}
+static void on_cloud_evt_data_received(const struct nrf_cloud_data *const data)
+{
+	sprintf(rsp_buf, "\r\n#XNRFCLOUD: %s\r\n", (char *)data->ptr);
```

Comment:
Need below condition be satisfied

if (nrf_cloud_ready) {

Reply:
i don't think that is needed anymore, but i will add it.
this is the handler for the new event `NRF_CLOUD_EVT_RX_DATA_GENERAL`; there should not be any of these events before the cloud is ready.
a new event `NRF_CLOUD_EVT_RX_DATA_SHADOW` may occur before the cloud is ready, but this data is ignored by slm.

Reply:
Yes it was the shadow data that confused customers before. Logically we should only allow uplink/downlink "general" after `NRF_CLOUD_EVT_READY`, so please add the `if` check even it's redundant now.